### ISSUE TITLE
Theme per tenant

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/OrchardCore#324]
@bleroy suggested this.

Have a Themes folder in per tenant folder to have custom themes that are only available for a tenant. This theme would probably not have any assembly so it should still be lightweight when used on hundreds of them. 

Advantage is that each tenant can have customizations, and hosters can grant access to this folder.

We could also do the same with Modules in order to deploy custom modules for tenants. We could also have Proxy modules that would just enable some first-level modules to be available in the tenant. It's a different way than having the list in the settings. Could still be a single configuration file.

It should be easy to implement with what @Jetski5822 is doing on extensions extensibility.

**Comment by sebastienros** [comment]
And maybe @jtkech is looking for something to do ... just saying.

**Comment by jtkech** [comment]
Oh, good ideas, so why not but i first need to well understand, let me a little time. Maybe step by step, e.g only begin with Themes per Tenant ...

Also, maybe good to 1st wait how shape table caching works. Before forgetting things, i'm finishing a 1st work related to the last discussion on this.

**Comment by Jetski5822** [comment]
I think the only thing that might need to be modified is ModuleService.cs.....

The lower lever stuff should not filter, only where we display and enable, which is done in the Module Service.

**Comment by Jetski5822** [comment]
Going to take a look once Extensions is in.

**Comment by sebastienros** [comment]
Some questions
- Should this theme be displayed in the list of themes in the admin? (easier if yes)
- Can it be disabled? (easier if yes)

So in the end it's just about loading another theme, but one that shouldn't be seen from other tenants, which is something I don't think we support today. We shouldn't duplicate the list of extensions, but maybe have a set of extension ids that are available for a tenant, and when we want to get the list of extensions, there is an option to set it for a named tenant, and then we only get these.

Should we then have a Modules folder and a Themes folder in the Sites/{Tenant} ?
What about extension name conflicts? We could totally have the same theme in different folders, or should we prefix their ids automatically with the tenant name, so that internally there can't be any conflict?

**Comment by Jetski5822** [comment]
Ah yeah - was thinking about these questions earlier.

- We will allow you to have what ever folder you want under the tenant. Configuration of the application will allow you to say that modules, themes etc are under app_data\sites\**\* ... something like that

- To avoid name clashes, we will prefix each extension with the tenant name, or maybe sub path..... sub path means that I can have the same extension id in two folders in the same tenant.

- Can it be disabled = yes

- the extension will be in everything we already see... I see Orchard loading them all up just like now... and applying filters to the UI to enable and disable.

- Maybe I need another Extension Provider... like a ShellExtensionProvider, which has a ShellExtensionInfo or something like that. Dont know yet.

**Comment by sebastienros** [comment]
What about just using a custom shape template harvester config to add this for a current tenant, After all we will mostly need to override shapes. That would solve most of the requirements for now. And to support Views then it would just be about working on the View Locations.

**Comment by sebastienros** [comment]
Well, apparently this is not the first time we've had this idea ;) https://github.com/OrchardCMS/Orchard2/issues/77

**Comment by jtkech** [comment]
So, a new TenantViewLocationExpander.

**Comment by Jetski5822** [comment]
No, the only thing we need is a new shell based extension provider. This will find extensions in shells.

Next modify shell feature provider to make sure correct ones shown.

Then nap.

**Comment by sebastienros** [comment]
@Jetski5822 think about it, handling it like a module will have some drawbacks:
- We'll have to deal with this module differently in terms of enabling/disabling
- This is be an extension visible to all tenants, so listing available modules will have to be handle specifically
- This requires more ceremony to setup as we need a manifest (and maybe a project file?)

Just adding a new ViewLocationProvider and an additional location in the ShapeTemplateHarvester would provide the same features, with less effort.
I would personally go with the simplest solution, as long as I didn't miss any obvious flaws. But if you have answers to my concerns I would definitely prefer to treat it as a standard theme.

**Comment by Jetski5822** [comment]
Okay so there are a number of issues in the extensions code for this to work, the main one being that we use the extension id in the paths to modules and themes... that wont work anymore - going to have a spike,.

**Comment by sebastienros** [comment]
Or if there are issues, keep it simple by just extending  views and shapes
resolution.

**Comment by Jetski5822** [comment]
There are, just writing tests at the minute to drive the implementation, will be great when it's working as it's highlighted some issues already

**Comment by sebastienros** [comment]
__WARNING__ (in bold) As I read all the changes and re-reading your comment:
> the only thing we need is a new shell based extension provider

This doesn't look like as easy as you state. Now, what about JUST adding a theme location expander and a shape template provider to just support theme overrides without having to deal with the extensions management specificities of these modules (code + UI)?

Please state the issue you are trying to solve. For me the issue is about being able to customize the theme per tenant.

**Comment by Jetski5822** [comment]
Yeah I was wrong, and that was quite apparent to me when I wrote my first test.

When I rewrote Extensions, I made an error in the api, because I didnt have a second implementation, that was to do with the manifests. Manifests say what your loading, and subpaths are where you are loading. In Orchard the two sit side by side, but that doesnt necessarily always have to be the case, and infact, by placing the manifest provider loader stuff inside the extension provider, you create a perf overhead too. Though fixing that was a side effect and not a goal.

What I wanted to do is to enable Extensions from anywhere easily, be them themes or modules. I found that by not using a wildcard solution you cant really do this easily.

So I guess the goal I tried to solve is to load extensions from anywhere, including shells to allow you to have Shell Themes... i.e. customise a theme per tenant
